ada_feeding_perception/ada_feeding_perception/food_on_fork.py

```diff
@@ -0,0 +1,363 @@
+            # frustum, then those pixels are converted to be 0
+            img_converted = np.where(
+                np.logical_or(cropped_img < self.min_dist, cropped_img > self.max_dist),
+                0,
+                1,
+            ).astype("uint8")
+            cropped_img_np = np.array(img_converted)
+
+            # self.get_logger().info(str(cropped_img_np))
+            # self.get_logger().info(str(np.count_nonzero(cropped_img_np == 1)))
+            # cv.imshow("cropped", cropped_img_np * 255)
+            # cv.waitKey(40)
+            X_test = cropped_img_np.flatten()
+
+            # Get the prediction, which is a float value
+            prediction = self.predict_food_on_fork(X_test=X_test)
+            # self.get_logger().info(str(prediction))
+            if self.test:
+                self.get_logger().info(str(prediction))
+            # ^^^^---ABOVE is the code when using EACH PIXEL as a feature---^^^^
+
+            float32msg = Float32()
+            float32msg.data = prediction
+            self.publisher_depth.publish(float32msg)
+
+    def food_on_fork_num_pixels(self, depth_img: np.ndarray) -> int:
+        """
+        Calculates the number of pixels in the provided depth image (through a depth image
+        message). This method is used to calculate the number of pixels that is used for the
+        Logistic Regression to output a confidence
+
+        Parameters:
+        ----------
+        depth_img: depth image
+        left_top_corner: Tuple(int, int): Top-left point of the bounding box rectangle
+        right_bottom_corner: Tuple(int, int): Bottom-right point of the bounding box of the
+            rectangle min_dist: int: minimum depth to consider (note
+            that 330 is approx distance to the fork tine)
+        max_dist: int: maximum depth to consider (note that 330 is approx distance to the fork tine)
+
+        Returns:
+        ----------
+        number of pixels within the specified parameter range
+        """
+
+        # consider the points for the rectangle
+        pt1_col, pt1_row = self.left_top_corner
+        pt2_col, pt2_row = self.right_bottom_corner
+
+        # create mask that satisfies the rectangle and distance conditions
+
+        # For instance, take a mask = [[F, F, F, F], [F, F, F, F], [F, F, F, F]]
+        mask_img = np.zeros_like(depth_img, dtype=bool)
+        # The pixels within the rectangular range will be true,
+        # resulting in mask = [[F, T, T, F], [F, T, T, F], [F, T, T, F]]
+        mask_img[pt1_row:pt2_row, pt1_col:pt2_col] = True
+        # The pixels within the depth range will be true,
+        # resulting in mask = [[F, F, F, F], [F, T, T, F], [F, F, F, F]]
+        mask_img[
+            np.logical_not((self.min_dist < depth_img) & (depth_img < self.max_dist))
+        ] = False
+
+        return np.count_nonzero(mask_img)
+
+    def predict_food_on_fork(
```

Comment:
We should do this conditioning on type of features only once. You already do it in `listener_callback_depth`. Then, this function should just take in `x: npt.NDArray` and pass that into the model and return the output. It shouldn't matter whether `x` has one feature or 10K. 

In general, you should make the assumption that the programmer specifically specified the rosparams (you can even check that in the load_params function if you want) and just pass in the datapoint with the num/type features they specified into the model. If they misspecified the rosparams, there will be an error, and they can address that.